packages/shylu/shylu_node/fastilu/src/shylu_fastilu.hpp

```diff
@@ -64,13 +64,172 @@
+  map_type proc_map(nrows, 0, comm_ptr, Tpetra::LocallyReplicated);
+  Teuchos::RCP<map_type> map_ptr =  Teuchos::RCP(&proc_map, false);
+  graph_type graph(map_ptr, map_ptr, rv, col_inds_host);
+  graph.fillComplete();
+  BCrsMatrix bcrs_matrix(graph, values_host, block_size);
+
+  auto A = Tpetra::convertToCrsMatrix(bcrs_matrix);
+  auto localA = A->getLocalMatrixDevice();
+  Kokkos::resize(row_map_host, localA.graph.row_map.size());
+  Kokkos::deep_copy(row_map_host, localA.graph.row_map);
+  Kokkos::resize(col_inds_host, localA.graph.entries.size());
+  Kokkos::deep_copy(col_inds_host, localA.graph.entries);
+  Kokkos::resize(values_host, localA.values.size());
+  Kokkos::deep_copy(values_host, localA.values);
+  unfill_crs(row_map_host, col_inds_host, values_host);
+}
+
+template <typename View1, typename View2, typename View3, typename View4>
+void reblock(View1& row_map, View2& row_idx, View3& col_inds, View4& values, const int block_size)
+{
+  using LO            = Tpetra::Map<>::local_ordinal_type;
+  using GO            = Tpetra::Map<>::global_ordinal_type;
+  using Node          = Tpetra::KokkosCompat::KokkosSerialWrapperNode;
+  using Scalar        = typename View4::non_const_value_type;
+  using TCrsMatrix    = Tpetra::CrsMatrix<Scalar, LO, GO, Node>;
+  using map_type      = typename TCrsMatrix::map_type;
+  using graph_type    = typename TCrsMatrix::crs_graph_type;
+  using row_vtype     = typename graph_type::local_graph_device_type::row_map_type::non_const_type;
+
+  row_vtype rv("temp", row_map.extent(0));
+  Kokkos::deep_copy(rv, row_map);
+  const auto nrows = row_map.size() - 1;
+  Teuchos::SerialComm<LO> SerialComm;
+  Teuchos::RCP<const Teuchos::Comm<LO> > comm_ptr = Teuchos::RCP(&SerialComm, false);
+  map_type proc_map(nrows, 0, comm_ptr, Tpetra::LocallyReplicated);
+  Teuchos::RCP<map_type> map_ptr =  Teuchos::RCP(&proc_map, false);
+  TCrsMatrix crs_matrix(map_ptr, map_ptr, rv, col_inds, values);
+  crs_matrix.fillComplete();
+  auto crs_matrix_block_filled = Tpetra::fillLogicalBlocks(crs_matrix, block_size);
+
+  auto bcrs_matrix = Tpetra::convertToBlockCrsMatrix(*crs_matrix_block_filled, block_size);
+  auto localA = bcrs_matrix->getLocalMatrixDevice();
+  auto rowptrs = localA.graph.row_map;
+  auto colinds = localA.graph.entries;
+  auto vals = localA.values;
+
+  Kokkos::resize(row_map, rowptrs.size());
+  Kokkos::deep_copy(row_map, rowptrs);
+  Kokkos::resize(col_inds, colinds.size());
+  Kokkos::deep_copy(col_inds, colinds);
+  Kokkos::resize(values, vals.size());
+  Kokkos::deep_copy(values, vals);
+
+  // Now do row idx
+  const auto nrows_blocked = row_map.size() -1;
+  Kokkos::resize(row_idx, colinds.size());
+  LO nnz = 0;
+  for (size_t row = 0; row < nrows_blocked; ++row) {
+    for (LO row_nnz = row_map(row); row_nnz < row_map(row+1); ++row_nnz) {
+      row_idx(nnz++) = row;
+    }
+  }
+  assert(nnz == row_map(nrows_blocked));
+}
+#endif
+
+// some useful preprocessor functions
+#ifdef FASTILU_TIMER
```

Comment:
Tried to encapsulate some things in macros to make the code a bit cleaner. This is unrelated to blockcrs.